test/vm.py

```diff
@@ -1,15 +1,47 @@
 import pathlib
 import subprocess
 import sys
+import uuid
 from io import StringIO
 
-from testutil import get_free_port, wait_ssh_ready
-
-
+import boto3
+from botocore.exceptions import ClientError
 from paramiko.client import AutoAddPolicy, SSHClient
 
+from testutil import AWS_REGION, get_free_port, wait_ssh_ready
+
 
 class VM:
+
+    def __del__(self):
+        self.force_stop()
+
+    def start(self):
+        raise NotImplementedError
+
+    def _log(self, msg):
+        # XXX: use a proper logger
+        sys.stdout.write(msg.rstrip("\n") + "\n")
+
+    def wait_ssh_ready(self):
+        wait_ssh_ready(self._ssh_port, sleep=1, max_wait_sec=600)
+
+    def force_stop(self):
+        raise NotImplementedError
+
+    def run(self, cmd, user, password):
+        raise NotImplementedError
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
```

Comment:
I think it would be nice if we could have a shared `run()` here as the ssh bits are a bit annoying to get right we can avoid some copy/paste this way :) We most likely need a common `running(), _ssh_port, _ip` between the classes.

Reply:
Done.